Add arg_iter lint to recommend IntoIterator over Iterator for argumen…

timsburk opened the pull request:
This PR implements the lint mentioned in issue #598  that suggests using `IntoIterator` instead of `Iterator` for function parameter types when possible.

## Implementation details
- Added a new lint in `examples/general/arg_iter` that detects when a function parameter type requires the `Iterator` trait
- The lint checks if the type parameter is used in other trait bounds (and avoids suggesting changes in those cases)
- Added documentation explaining why using `IntoIterator` provides more flexibility
- Implemented UI tests to verify the lint catches appropriate cases

## Testing
- Added test cases covering correct detection of `Iterator` bounds that could be changed
- Added tests for cases where the lint should not trigger (when `Iterator` is needed for other bounds)
- Verified the error messages provide helpful suggestions for fixing the issue

This allows functions to accept a wider range of types (like slices, arrays, vectors) without requiring callers to explicitly call `.iter()` or `.into_iter()`.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trailofbits/dylint?pullRequest=1581) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/trailofbits/dylint?pullRequest=1581) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trailofbits/dylint?pullRequest=1581) it.</sub>

timsburk commented:
Hello @smoelius, I have made a PR for issue mentioned in #598. Can you please review and give feedbacks?

smoelius commented:
@timsburk Sorry for the delay. I have a little bit of a reviewing backlog. But I will try to get to this soon.

smoelius commented:
@timsburk Sorry I haven't had a chance to review this. I am getting ready to travel, and I will get back early next week. I will make it a priority to review this then.

smoelius reviewed (commented):
@timsburk Thanks very much for your patience.

Let's please move this lint from `general` to `supplementary`.

As noted below, I think the current tests are not being run. I will need to review the lint's logic more carefully, but I would like to see the tests run first.

timsburk commented:
@smoelius made the changes you requested:

- Moved the lint from `general` to `supplementary`
- Added `ui_test`
- Added `dylint_linting::declare_late_lint!`

Please review and let me know about it.

Thank You!

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,205 @@
+
+        // Check function inputs (arguments)
+        // We need to correlate FnDecl inputs with the GenericParams to find the names
+        let input_tys_and_param_defs: Vec<_> = _decl.inputs
```
I would not use `_decl` here, because it is `HIR`.

The lint will be more robust if you use `rustc_middle` types.

You can use `cx.tcx.fn_sig(id)` to get the function's [`FnSig`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/type.FnSig.html). You can then use its [`inputs`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/type.FnSig.html#method.inputs) method to get the input types.

Rather than start with the parameters and try to map them to inputs, I would start with the inputs.

In particular, I would look for a [`Param`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/type.TyKind.html#variant.Param) type.

Note that the `Param`'s [`ParamTy`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/struct.ParamTy.html) has an index into the function's generic arguments.

So, if you start with `rustc_middle` input types, you should be able to map them relatively easily to generics.

I think this will be easier than starting with the generics and trying to map them back to the inputs.

smoelius reviewed (commented):
@timsburk I'm going to need a few iterations to review this. I hope that's okay.

I left some comments below. Please let me know if they are not specific enough.

timsburk commented:
Thank You @smoelius for the review.

> I'm still curious about this. :)

So apparently there is a bug that kept warning me about the absolute home path when I ran `cargo test`. It looks like someone else is already working on this issue in #1604. Adding that line temporarily stopped that warning. It is something I forgot to remove while pushing the branch, sorry for creating the confusion.

And for the other changes, I'll try to get things done soon.

Thanks again!

smoelius commented:
> It is something I forgot to remove while pushing the branch, sorry for creating the confusion.
> 
> And for the other changes, I'll try to get things done soon.
> 
> Thanks again!

No worries! Thank you!

timsburk commented:
@smoelius added `rustc_middle` types as you requested

timsburk commented:
Hi @smoelius,

Sorry to bother you, but could you please help me a but to get the CI to pass? I've tried a few times, but I couldn't get the `inconsistent qualification` and `fmt` ones to pass.

Thanks a lot!

smoelius commented:
It's no bother.

The `fmt` test wants to swap these two lines: https://github.com/trailofbits/dylint/blob/6442f1bd9c99e42b2e8f8d8d22ce72886db41655/examples/supplementary/arg_iter/src/lib.rs#L9-L10
I'm not sure if that's an edition 2024 thing. But either way, if you replace the lines with the following, the problem should go away:
```rust
use rustc_hir::{FnDecl, intravisit::FnKind};
```
The `inconsistent_qualification` lint wants all `ty` items to be imported, or to be referred to with a `ty::` prefix, but not both. Currently, both types of uses occur, e.g.:
- https://github.com/trailofbits/dylint/blob/6442f1bd9c99e42b2e8f8d8d22ce72886db41655/examples/supplementary/arg_iter/src/lib.rs#L105
- https://github.com/trailofbits/dylint/blob/6442f1bd9c99e42b2e8f8d8d22ce72886db41655/examples/supplementary/arg_iter/src/lib.rs#L56

Because `ty::` is only four characters, I often just use the prefix.

timsburk commented:
Thank you so much @smoelius. I fixed those things, hopefully it will pass now 🤞

timsburk commented:
@smoelius CI passed!

smoelius commented:
I'll look at this as soon as I can. (It will probably take me a few iterations to get through this, though.)

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+        let predicates = cx.tcx.predicates_of(id);
+
+        // Find generic type parameters that are *only* bound by Iterator (+ implicit Sized)
+        let mut iterator_only_params: Vec<Symbol> = Vec::new(); // Store only the name (Symbol)
+        for param in &generics.own_params {
+            // Iterate over a slice
+            if let ty::GenericParamDefKind::Type { .. } = param.kind {
+                // param.def_id is already the DefId
+                if is_param_bound_only_by_iterator(cx, predicates, param.def_id, iterator_def_id) {
+                    iterator_only_params.push(param.name); // Push only the name
+                }
+            }
+        }
```
I think I was not sufficiently clear. I think it should be possible to eliminate `iterator_only_params` and this loop. (See below.)

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+fn is_param_bound_only_by_iterator<'tcx>(
+    cx: &LateContext<'tcx>,
+    predicates: ty::GenericPredicates<'tcx>,
+    param_def_id: DefId, // The DefId of the type parameter we are checking (e.g., `I`)
```
I think it should be possible to pass a `rustc_middle::ty::Ty` here, rather than a `DefId`.

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+            let self_ty = bound_trait_pred.self_ty();
+
+            // Check if the bound applies to our parameter `param_def_id`
+            if let ty::TyKind::Param(p_ty) = self_ty.kind() {
```
Here, I think you should be able to just compare the passed-in `Ty` to `self_ty`.

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+
+            // Also check if the parameter appears in the *arguments* of another trait bound
+            // Example: `fn foo<T, U: SomeTrait<T>>(arg: U)` - here T appears in U's bound.
+            for arg in bound_trait_pred.trait_ref.args.iter() {
```
Could you move this check into a separate function?

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,192 @@
+                let param_def = generics.param_at(param_ty.index as usize, cx.tcx);
+
+                // Check if this parameter's name is in our list of iterator-only params
+                if iterator_only_params.contains(&param_def.name) {
```
Here, I think it should be possible to call `is_param_bound_only_by_iterator` directly, eliminating the need to call it and store its results ahead of time.

Please note that I'm recommending some changes to that function below.

smoelius reviewed (commented):
I feel silly asking this, but would it be possible to reduce the comments to just the ones you consider essential?

I know more comments are generally better than too few. But I feel that many of the comments here are a little too tangential or obvious, and therefore kind of distracting.

Thanks very much for your work on this.

timsburk commented:
Hello @smoelius — I've made the changes you requested.

However, the CI test was failing, so I looked into the logs and saw an `Error 429`, which indicates too many requests. It seems the CI might be rate-limited due to high activity, possibly from multiple PRs running simultaneously.

Just wanted to flag this in case it needs attention.

smoelius commented:
> Hello @smoelius — I've made the changes you requested.
> 
> However, the CI test was failing, so I looked into the logs and saw an `Error 429`, which indicates too many requests. It seems the CI might be rate-limited due to high activity, possibly from multiple PRs running simultaneously.
> 
> Just wanted to flag this in case it needs attention.

Curious. Thanks for the heads up.

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+                if current_trait_def_id == iterator_def_id {
+                    found_iterator_bound = true;
+                } else if current_trait_def_id != sized_def_id {
+                    found_other_bound = true;
+                    break;
```
```suggestion
                    return false;
```
I think if you've found another bound, you can return early.

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    if found_iterator_bound
+        && !found_other_bound
+        && is_param_used_in_other_trait_args(
+            cx,
+            predicates,
+            param_ty,
+            iterator_def_id,
+            sized_def_id,
+            fn_id,
+        )
+    {
+        found_other_bound = true;
+    }
+
+    found_iterator_bound && !found_other_bound
```
```suggestion
    found_iterator_bound
```
I would call `is_param_used_in_other_trait_args` separately.

Calling `is_param_used_in_other_trait_args` from `is_param_bound_only_by_iterator` doesn't seem intuitive to me.

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+                    iterator_def_id,
+                    sized_def_id,
+                    id,
+                ) {
```
```suggestion
                ) && !is_param_used_in_other_trait_args(
                    cx,
                    predicates,
                    *input_ty_skipped_binder,
                    iterator_def_id,
                    sized_def_id,
                    id,
                ) {
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+
+/// Checks if a given type parameter `param_ty` is bound _only_ by `iterator_def_id`
+/// within the given `predicates`. Implicit `Sized` bounds (`sized_def_id`) are ignored.
+/// Also checks if the parameter is used in the arguments of other trait bounds.
```
```suggestion
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+    fn_id: LocalDefId,
+) -> bool {
+    let mut found_iterator_bound = false;
+    let mut found_other_bound = false;
```
```suggestion
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+
+            for arg in bound_trait_pred.trait_ref.args.iter() {
+                if let Some(ty_arg) = arg.as_type() {
+                    if let ty::TyKind::Param(p_ty_arg) = ty_arg.kind() {
```
Can you just compare `p_ty_arg` and `param_ty` here, rather than pull out `param_ty`'s def id?

smoelius reviewed (commented):
This is looking pretty good. I think we're almost there.

timsburk commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+            return;
+        }
+
+        // Check function inputs using rustc_middle types for robustness
```
I initially thought about using 'simpler,' but then I felt 'robust' sounded more like something programmers would say, so I went with 'robust' instead 😅

timsburk commented:
@smoelius added the changes you mentioned. Please check.
I've also rebased on master branch

timsburk commented:
@smoelius just a quick update — I've rebased this branch onto master, and it looks like all the latest changes from master are now here. I've also squashed all the commits as there were a lot of them. Let me know if everything looks good on your end or if anything needs adjusting.

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,164 @@
+    def_id::{DefId, LocalDefId},
+};
+
+dylint_linting::declare_late_lint! {
+    pub ARG_ITER,
```
Could you please add doc comments here to match your README?

A good example can be seen in the recent PR #1623: https://github.com/trailofbits/dylint/pull/1623/files#diff-910f3f4e06c9a9d6a262f36213e71630d42f1749c01f2b6d979596fabef2fcfcR18-R93

Note that there is a script [update_example_READMEs.sh](https://github.com/trailofbits/dylint/blob/master/scripts/update_example_READMEs.sh) that generates READMEs from the examples' doc comments. Ideally, the doc comments you add would be such that your existing README is generated from them.

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,164 @@
+/// Checks if a given type parameter `param_ty` is bound _only_ by `iterator_def_id`
+/// within the given `predicates`
+fn is_param_bound_only_by_iterator<'tcx>(
+    _cx: &LateContext<'tcx>,
```
```suggestion
```
For functions known only to this file, could you please remove all unnecessary arguments?

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,164 @@
+                        None,
+                        format!(
+                            "consider using `IntoIterator` instead of `Iterator` for parameter `{}`",
+                            param_def.name
```
```suggestion
                            param_ty.name
```
I think you can get the parameter [name](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/struct.ParamTy.html#structfield.name) from the `ParamTy`, making the use of generics unnecessary.

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,164 @@
+            return;
+        };
+
+        let generics = cx.tcx.generics_of(id);
```
```suggestion
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,164 @@
+        let generics = cx.tcx.generics_of(id);
+        let predicates = cx.tcx.predicates_of(id);
+
+        if generics.own_params.is_empty() {
+            return;
+        }
```
```suggestion
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,164 @@
+        for (i, input_ty) in fn_sig.inputs().iter().enumerate() {
+            let input_ty_skipped_binder: &ty::Ty = input_ty.skip_binder(); // Skip binder once
+            if let ty::TyKind::Param(param_ty) = input_ty_skipped_binder.kind() {
+                let param_def = generics.param_at(param_ty.index as usize, cx.tcx);
```
```suggestion
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,164 @@
+            }
+
+            for arg in bound_trait_pred.trait_ref.args.iter() {
+                if let Some(ty_arg) = arg.as_type() {
+                    if let ty::TyKind::Param(_) = ty_arg.kind() {
+                        if ty_arg == param_ty {
+                            return true;
+                        }
+                    }
+                }
```
```suggestion
                if arg.as_type() == Some(param_ty) {
                    return true;
                }
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,164 @@
+    sized_def_id: DefId,
+    _fn_id: LocalDefId,
+) -> bool {
+    if !matches!(param_ty.kind(), ty::TyKind::Param(_)) {
+        return false;
+    }
```
```suggestion
```
Unless you feel strongly, I would drop this. It's verified that `param_ty` is a `Parm` just before `is_param_used_in_other_trait_args` is called. I don't think it's necessary to check it again here.

smoelius reviewed (commented):
I _think_ this should be my last round of comments.

Thanks a lot for this! This is looking really good!

timsburk commented:
Hi @smoelius,
I've made the changes you requested. However, I am not entirely sure about the README and doc comments part, and I suspect I may have implemented that part incorrectly. I believe that's why some of the CI tests are currently failing.

smoelius commented on examples/supplementary/arg_iter/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "arg_iter"
+version = "4.1.0"
+edition = "2021"
```
```suggestion
edition = "2024"
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,181 @@
+
+dylint_linting::declare_late_lint! {
+    /// ### What it does
+    /// 
```
```suggestion
    ///
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,181 @@
+    /// `IntoIterator` instead.
+    ///
+    /// ### Why is this bad?
+    /// 
```
```suggestion
    ///
```

smoelius commented on examples/supplementary/arg_iter/src/lib.rs:
```diff
@@ -0,0 +1,181 @@
+    /// `.iter()` calls. This often makes the API easier to use.
+    ///
+    /// ### Example
+    /// 
```
```suggestion
    ///
```

smoelius reviewed (commented):
I think the failures are because of edition 2021 and some spaces after `///`.

timsburk commented:
@smoelius, I managed to pass the CI after fixing the docs and README. There was one more CI fail, in [`27a5897`](https://github.com/trailofbits/dylint/pull/1581/commits/27a589717dc439a19f3f8a0600496de9aae5653d) about this change in `utils/testing/src/lib.rs`,

```rust
let _ = ["fixed", "stderr", "stdout"]
        .map(|extension| copy_with_extension(&target.src_path, &to, extension).unwrap_or_default());
``` 

I am not entirely sure about this, the CI logs suggested this change.

smoelius commented:
@timsburk Thanks for the heads up. However, I would prefer to address that warning in a more type-explicit way. I am currently testing #1644.

Apologies. I thought I sent this message around the time I opened #1644, but it appears I didn't.

smoelius commented:
#1644 is merged. If you could, please:
- undo commit aaaec13b00d136d5e217d60bcc03c0e3de72590c
- rebase onto master
- squash your commits down to one

Then I think this is good to go.

timsburk commented:
@smoelius done ✅

smoelius commented:
The failure looks sporadic. Let's try again.
